# Importing libraries

In [58]:
import pandas as pd
import numpy as np
import os

# Path

In [59]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'C:\Users\Mallika\Documents\Career Foundry\Module2- Data Immersion\Achievement 4\June2025_Instacart Basket Analysis'

# Importing Data

In [60]:
# Importing ords_prods_merge data
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

# 1.Creating a subset

# 1.1 Create "price_label" column

In [107]:
def price_label(row):
    if row['prices'] <= 5: return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15): return 'Mid-range product'
    elif row['prices'] > 15:  return 'High range product'
    else: return 'Not enough data'

In [62]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Mallika\AppData\Local\Temp\ipykernel_23168\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [63]:
# Creating a subset due to the large size of the dataframe- first 1 million rows
df = ords_prods_merge[:1000000]

In [64]:
df.shape

(1000000, 15)

In [65]:
# Use loc() method on df subset to apply price range function.
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

C:\Users\Mallika\AppData\Local\Temp\ipykernel_23168\3848246037.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [66]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     673340
Low-range product     314392
High-range product     12268
Name: count, dtype: int64

In [71]:
df_ords_prods_merge.apply(price_label, axis=1)

0           Mid-range product
1           Mid-range product
2           Low-range product
3           Low-range product
4           Low-range product
                  ...        
32404854    Mid-range product
32404855    Mid-range product
32404856    Low-range product
32404857    Low-range product
32404858    Low-range product
Length: 32404859, dtype: object

In [69]:
# Use loc() method on entire ords_prods_merge dataframe to apply price range function.
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [72]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

In [93]:
df_ords_prods_merge

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0          2539329        1             1          2                  8   
1          2539329        1             1          2                  8   
2          2539329        1             1          2                  8   
3          2539329        1             1          2                  8   
4          2539329        1             1          2                  8   
...            ...      ...           ...        ...                ...   
32404854   2977660   206209            13          1                 12   
32404855   2977660   206209            13          1                 12   
32404856   2977660   206209            13          1                 12   
32404857   2977660   206209            13          1                 12   
32404858   2977660   206209            13          1                 12   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                            NaN         196                  1          0   
1                            NaN       14084                  2          0   
2                            NaN       12427                  3          0   
3                            NaN       26088                  4          0   
4                            NaN       26405                  5          0   
...                          ...         ...                ...        ...   
32404854                     7.0       14197                  5          1   
32404855                     7.0       38730                  6          0   
32404856                     7.0       31477                  7          0   
32404857                     7.0        6567                  8          0   
32404858                     7.0       22920                  9          0   

         _merge                             product_name  aisle_id  \
0          both                                     Soda        77   
1          both  Organic Unsweetened Vanilla Almond Milk        91   
2          both                      Original Beef Jerky        23   
3          both               Aged White Cheddar Popcorn        23   
4          both         XL Pick-A-Size Paper Towel Rolls        54   
...         ...                                      ...       ...   
32404854   both                             Tomato Paste         9   
32404855   both          Brownie Crunch High Protein Bar         3   
32404856   both    High Protein Bar Chunky Peanut Butter         3   
32404857   both      Chocolate Peanut Butter Protein Bar         3   
32404858   both      Roasted & Salted Shelled Pistachios       117   

          department_id  prices match    price_range_loc  
0                     7     9.0  both  Mid-range product  
1                    16    12.5  both  Mid-range product  
2                    19     4.4  both  Low-range product  
3                    19     4.7  both  Low-range product  
4                    17     1.0  both  Low-range product  
...                 ...     ...   ...                ...  
32404854              9     5.6  both  Mid-range product  
32404855             19     5.9  both  Mid-range product  
32404856             19     4.2  both  Low-range product  
32404857             19     4.9  both  Low-range product  
32404858             19     1.7  both  Low-range product  

[32404859 rows x 16 columns]

# 1.2. Create"busiest_day"

In [106]:
# Find frequency of orders_day_of_the_week.
df_ords_prods_merge['order_dow'].value_counts(dropna = False)

order_dow
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [96]:
# Use for-loop to determine whether orders are on \"busiest day\" (0 = Saturday), \"least busy\" (4 = Wednesday), or \"regularly busy\" (other days of the week).\n",
result = []
for value in df_ords_prods_merge["order_dow"]:
    if value == 0: result.append("Busiest day")
    elif value == 4: result.append("Least busy")
    else: result.append("Regularly busy")

In [97]:
# Create new column in ords_prods_merge to view results in context.
df_ords_prods_merge['busiest_day'] = result

In [98]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 2. Update"busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."

In [102]:
# Create column "busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days).
result_2 = []
for value in df_ords_prods_merge["order_dow"]:
    if value == 0 or value == 1:  result_2.append("Busiest days")
    elif value == 4 or value == 3: result_2.append("Slowest days")
    else: result_2.append("Regularly busy")

# 3. Check values of new "busiest days" column for accuracy with observations in markdown format.

In [103]:
# Create new 'busiest_days' column in ords_prods_merge to view results in context.
df_ords_prods_merge['busiest_days'] = result_2

In [104]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [105]:
# Check output of ords_prods_merge with new "busiest days" column.
df_ords_prods_merge.head ()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    price_range_loc     busiest_day    busiest_days  
0  both  Mid-range product  Regularly busy  Regularly busy  
1  both  Mid-range product  Regularly busy  Regularly busy  
2  both  Low-range product  Regularly busy  Regularly busy  
3  both  Low-range product  Regularly busy  Regularly busy  
4  both  Low-range product  Regularly busy  Regularly busy

The total value counts for the ords_prods_merge dataframe is equal to the total sum of all groupings (i.e. "Regularly busy" + "Busiest days" + "Slowest days"). In addition, the listed sums for each label match the sum totals for the qualifying days of the week (e.g. Total "Busiest days" = 11864412, which matches order_dow 0 + 1, or 6204182 + 5660230).

# 4. Create new column "busiest_period_of_day" to identify time labeled “Most orders,” “Average orders,” and “Fewest orders.”

In [108]:
# Check value counts in "order_hour_of_day" column
df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16, and 9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, and 3; "Average orders" = all remaining values in order_hour_of_day.

In [109]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”
result_3 = []
for value in df_ords_prods_merge["order_hour_of_day"]:
    if value in [10, 11, 14, 15, 13, 12, 16, 9]: result_3.append("Most orders")
    elif value in [23, 6, 0, 1, 5, 2, 4, 3]: result_3.append("Fewest orders")
    else: result_3.append("Average orders")

In [110]:
# Create new column "busiest_period_of_day" in ords_prods_merge
df_ords_prods_merge['busiest_period_of_day'] = result_3

# 5. Print frequency of "busiest_period_of_day" column


In [111]:
# Print value counts in \"busiest_period_of_day\" column
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [112]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 19 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   order_dow               int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   _merge                  category
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  match                   category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
dtypes: category(2), float64(2), int64(10), object(5)
memory usage: 4.2+ GB


# Export new dataframe as a pickle file 

In [113]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))